pink/rondis/string/db_operations.cc

```diff
@@ -614,3 +614,173 @@ int rondb_get_rondb_key(const NdbDictionary::Table *tab,
     }
     return 0;
 }
+
+#define RONDB_INSERT 2
+#define RONDB_UPDATE 1
+#define REG0 0
+#define REG1 1
+#define REG2 2
+#define REG3 3
+#define REG4 4
+#define REG5 5
+#define REG6 6
+#define REG7 7
+#define LABEL0 0
+#define LABEL1 1
+void incr_key_row(std::string *response,
+                  Ndb *ndb,
+                  const NdbDictionary::Table *tab,
+                  NdbTransaction *trans,
+                  struct key_table *key_row) {
+
+    const NdbDictionary::Column *value_start_col = tab->getColumn(KEY_TABLE_COL_value_start);
+    const NdbDictionary::Column *tot_value_len_col = tab->getColumn(KEY_TABLE_COL_tot_value_len);
+    const NdbDictionary::Column *rondb_key_col = tab->getColumn(KEY_TABLE_COL_rondb_key);
+
+    NdbOperation::OperationOptions opts;
+    std::memset(&opts, 0, sizeof(opts));
+    /**
+     * The mask specifies which columns is to be updated after the interpreter
+     * has finished. The values are set in the key_row.
+     * We have 7 columns, we will update tot_value_len in interpreter, same with
+     * value_start.
+     *
+     * The rest, redis_key, rondb_key, value_data_type, num_rows and expiry_date
+     * are updated through final update.
+     */
+
+    const Uint32 mask = 0x55;
+    const unsigned char *mask_ptr = (const unsigned char *)&mask;
+
+    // redis_key already set as this is the Primary key
+    key_row->null_bits = 1; // Set rondb_key to NULL, first NULL column
+    key_row->num_rows = 0;
+    key_row->value_data_type = 0;
+    key_row->expiry_date = 0;
+
+    /* Define the interpreted program */
+    Uint32 code_buffer[128];
+    NdbInterpretedCode code(tab, &code_buffer[0], sizeof(code_buffer));
+    code.load_const_u16(REG0, 4); //Memory offset 0
+    code.load_const_u16(REG6, 0); //Memory offset 0
+    int ret_code = code.load_op_type(REG1); // Read operation type into register 1
+    code.branch_eq_const(REG1, RONDB_INSERT, LABEL1); //Inserts go to label 1
+
+    /**
+     * The first 4 bytes of the memory must be kept for the Attribute header
+     * REG0 Memory offset == 4
+     * REG1 Memory offset == 6
+     * REG2 Size of value_start
+     * REG3 Size of value_start without length bytes
+     * REG4 Old integer value after conversion
+     * REG5 New integer value after increment
+     * REG6 Memory offset == 0
+     * REG7 Value of rondb_key (should be NULL)
+     */
+    /* UPDATE code */
+    code.read_attr(REG7, rondb_key_col);
+    code.branch_eq_null(REG7, LABEL0);
+    code.interpret_exit_nok();
```

Comment:
`_nok`?

Reply:
nok is the opposite of ok (nok = short for Not OK). It will abort the operation and report a failure to the application.

Reply:
Here is the docs of this call:
  /* interpret_exit_nok
   *
   * Scanning operation     : Error codes 626 and 899: This row should not be
   *                          returned as part of the scan.  Move onto next row.
   *                          Error codes [6000-6999]: Abort the scan.
   *
   * Non-scanning operation : Abort the operation
   *
   * Space required        Buffer    Request message
   *   interpret_exit_nok  1 word    1 word
   *
   * @param ErrorCode An error code which will be returned as part
   * of the operation.  If not supplied, defaults to 626. Applications should
   * use error code 626 or any code in the [6000-6999] range. Error code 899
   * is supported for backwards compatibility, but 626 is recommended instead.
   * For other codes, the behavior is undefined and may change at any time
   * without prior notice.
   *
   * @return 0 if successful, -1 otherwise
   */
  int interpret_exit_nok(Uint32 ErrorCode);
  int interpret_exit_nok();